// joi-password/lib/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.joiPasswordExtendCore = void 0;
function joiPasswordExtendCore(joi) {
    return {
        type: 'string',
        base: joi.string(),
        messages: {
            'password.minOfUppercase': '{#label} should contain at least {#min} uppercase character',
            'password.minOfSpecialCharacters': '{#label} should contain at least {#min} special character',
            'password.minOfLowercase': '{#label} should contain at least {#min} lowercase character',
            'password.minOfNumeric': '{#label} should contain at least {#min} numeric character',
            'password.noWhiteSpaces': '{#label} should not contain white spaces',
            'password.onlyLatinCharacters': '{#label} should only contain latin characters',
            'password.doesNotInclude': '{#label} should not contain any of the following: {#patterns}',
        },
        rules: {
            minOfUppercase: {
                method(min) {
                    return this.$_addRule({
                        name: 'minOfUppercase',
                        args: { min },
                    });
                },
                args: [
                    {
                        name: 'min',
                        assert: (value) => typeof value === 'number' && !isNaN(value),
                        message: 'must be a number',
                    },
                ],
                validate: (value, helpers, { min = 0 }) => {
                    if (!new RegExp(`(?=(.*[A-Z]){${min}})`).test(value))
                        return helpers.error('password.minOfUppercase', { min });
                    return value;
                },
            },
            minOfLowercase: {
                method(min) {
                    return this.$_addRule({
                        name: 'minOfLowercase',
                        args: { min },
                    });
                },
                args: [
                    {
                        name: 'min',
                        assert: (value) => typeof value === 'number' && !isNaN(value),
                        message: 'must be a number',
                    },
                ],
                validate: (value, helpers, { min = 0 }) => {
                    if (!new RegExp(`(?=(.*[a-z]){${min}})`).test(value))
                        return helpers.error('password.minOfLowercase', { min });
                    return value;
                },
            },
            minOfSpecialCharacters: {
                method(min) {
                    return this.$_addRule({
                        name: 'minOfSpecialCharacters',
                        args: { min },
                    });
                },
                args: [
                    {
                        name: 'min',
                        assert: (value) => typeof value === 'number' && !isNaN(value),
                        message: 'must be a number',
                    },
                ],
                validate: (value, helpers, { min = 0 }) => {
                    const numSpecial = value.length - (value.match(/[a-zA-Z0-9]/g) || []).length;
                    if (numSpecial < min)
                        return helpers.error('password.minOfSpecialCharacters', {
                            min,
                        });
                    return value;
                },
            },
            minOfNumeric: {
                method(min) {
                    return this.$_addRule({
                        name: 'minOfNumeric',
                        args: { min },
                    });
                },
                args: [
                    {
                        name: 'min',
                        assert: (value) => typeof value === 'number' && !isNaN(value),
                        message: 'must be a number',
                    },
                ],
                validate: (value, helpers, { min = 0 }) => {
                    if (!new RegExp(`(?=(.*[0-9]){${min}})`, 'g').test(value))
                        return helpers.error('password.minOfNumeric', { min });
                    return value;
                },
            },
            noWhiteSpaces: {
                validate: (value, helpers) => {
                    if (new RegExp(` `, 'g').test(value))
                        return helpers.error('password.noWhiteSpaces');
                    return value;
                },
            },
            onlyLatinCharacters: {
                validate: (value, helpers) => {
                    if (new RegExp(`[^a-zA-Z0-9!@#$%^&*()_+\\-=\\[\\]{};':"\\\\|,.<>\\/? ]`, 'g').test(value))
                        return helpers.error('password.onlyLatinCharacters');
                    return value;
                },
            },
            doesNotInclude: {
                method(patterns) {
                    return this.$_addRule({
                        name: 'doesNotInclude',
                        args: { patterns },
                    });
                },
                args: [
                    {
                        name: 'patterns',
                        assert: (value) => Array.isArray(value),
                        message: 'must be an array',
                    },
                ],
                validate: (value, helpers, { patterns = [] }) => {
                    const matches = patterns.filter((pattern) => new RegExp(pattern, 'g').test(value));
                    if (matches.length > 0)
                        return helpers.error('password.doesNotInclude', { patterns: matches.join(', ') });
                    return value;
                },
            },
        },
    };
}
exports.joiPasswordExtendCore = joiPasswordExtendCore;
